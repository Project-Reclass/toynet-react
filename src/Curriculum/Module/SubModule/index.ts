import SubModule from './SubModule';
export default SubModule;