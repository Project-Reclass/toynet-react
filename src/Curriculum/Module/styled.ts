/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
import styled from '@emotion/styled';
import { Icon } from '@chakra-ui/react';
import { TriangleUpIcon } from '@chakra-ui/icons';

interface RotatableArrowProps {
  rotated: boolean;
}

export const RotatableIcon = styled(Icon)`
  transition: 0.3s ease;
  cursor: pointer;

  transform: ${({ rotated }: RotatableArrowProps) => rotated && 'rotate(90deg)'};
`;

export const RotatableTriangle = styled(TriangleUpIcon)`
  transition: 0.3s ease;
  cursor: pointer;

  transform: ${({ rotated }: RotatableArrowProps) => rotated && 'rotate(90deg)'};
`;
