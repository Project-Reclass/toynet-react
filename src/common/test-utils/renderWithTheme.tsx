import React, { FC } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '@chakra-ui/core/dist';
import renderer from 'react-test-renderer';

const Wrapper: FC = ({ children }) => (
  <ThemeProvider>{children}</ThemeProvider>
);

export const renderWithTheme = (ui: any) => {
  return render(ui, { wrapper: Wrapper });
};

export const renderTreeWithTheme = (ui: any) => {
  return renderer.create(<Wrapper>{ui}</Wrapper>);
};
