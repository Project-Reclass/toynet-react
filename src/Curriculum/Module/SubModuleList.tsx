/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.  */
import { Flex } from '@chakra-ui/core';
import React, { FC } from 'react';

import { ModuleInterface } from './types';
import SubModule from './SubModule/SubModule';

interface Props {
  subModules: ModuleInterface[];
}

const SubModuleList: FC<Props> = ({ subModules }) => (
  <Flex flexDirection='column' width='80%' marginLeft='auto'>
    {subModules.map((module, idx) => (
      <SubModule
        {...module}
        index={idx}
        count={subModules.length}
        key={`${module.id}-${module.title}-${module.moduleId}`}
      />
    ))}
  </Flex>
);

export default SubModuleList;