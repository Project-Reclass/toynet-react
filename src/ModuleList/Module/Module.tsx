import React, { FC, useState } from 'react';
import { ProgressBar } from 'react-bootstrap';

import { ReactComponent as CaretIcon } from 'src/assets/caret-up-solid.svg';

import styles from './Module.module.css';
import SubModule from './SubModule';

export enum ModuleTypes {
  QUIZ = 'quiz',
  ARTICLE = 'article',
  PARENT = 'parent',
  EMULATOR = 'emulator',
}

export interface ModuleInterface {
  id: number | string;
  moduleId: number | string;
  title: string;
  progress: number;
  type: ModuleTypes;
}

interface Props {
  subModules: ModuleInterface[];
}

const SubModules: FC<Props> = ({ subModules }) => (
  <div className={styles.dropdownItemContainer}>
    {subModules.map(module => (
      <SubModule
        {...module}
        key={`${module.id}-${module.title}-${module.moduleId}`}
      />
    ))}
  </div>
);

const Module: FC<Props & ModuleInterface> = ({ title, progress, subModules }) => {
  const [showSubModules, setShowSubModules] = useState(true);

  return (
    <div>
      <div className={styles.dropdownContainer}>
        <span
          data-testid="caret"
          className={`${styles.arrow} ${showSubModules && styles.rotated}`}
          onClick={() => setShowSubModules(!showSubModules)}
        >
          <CaretIcon />
        </span>
        <span>
          {title}
        </span>
        <span>
          <ProgressBar label={`${progress}%`} now={progress} />
        </span>
      </div>
      {showSubModules && <SubModules subModules={subModules} />}
    </div>
  );
};

export default Module;