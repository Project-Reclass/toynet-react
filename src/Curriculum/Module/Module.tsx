/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
import React, { FC } from 'react';
import { Box, Collapse, Flex, Text, Tooltip } from '@chakra-ui/core';
import { ModuleIntf } from 'src/common/types/curriculum';

import { RotatableIcon } from './styled';
import SubModuleList from './SubModuleList';
import { useSessionStorage } from 'src/common/hooks/useSessionStorage';

interface Props {
  index: number;
  locked: boolean;
  paddingTop?: string;
}

const withToolTip = (Component: React.ReactNode) => (
  <Tooltip
    hasArrow
    label='ðŸ”’ This module is currently locked.'
    {...{'aria-label': 'This module is currently locked.'}}
  >
    {Component}
  </Tooltip>
);

/**
 * Wraps a components with a tooltip that is only shown if `isLocked` is true
 */
const TooltipIslocked: React.FC<{isLocked: boolean}> = ({children, isLocked}) => {
  if (isLocked)
    return withToolTip(children);

  return <>{children}</>;
};

const Module: FC<ModuleIntf & Props> = (
  {
    id,
    index,
    locked,
    introduction,
    name,
    submodules,
    paddingTop,
  },
) => {
  const [isOpen, setOpen] =
    useSessionStorage<boolean>(`module-${id}-${index}`, false,
      value => JSON.parse(value));

  return (
    <Box
    paddingTop={paddingTop}
    borderLeft={'2pt solid white'}
    >
      <Flex onClick={() => setOpen(open => !open)} cursor='pointer'>
        <RotatableIcon
          name={'triangle-up'}
          rotated={isOpen}
          size='1.5rem'
          marginY='auto'
          marginX='1rem'
        />
        <Flex justifyContent='space-between' width='100%'>
          <TooltipIslocked isLocked={locked}>
            <Text fontSize='2xl' userSelect='none'>
              {`Module ${index + 1}: ${name}`}
            </Text>
          </TooltipIslocked>
          <Text fontSize='1xl' userSelect='none' m='auto 0'>
            {`${submodules.length} / ${submodules.length} completed, ${submodules.length} in progress`}
          </Text>
        </Flex>
      </Flex>
      <Collapse isOpen={isOpen}>
        <Text fontSize='1xl' userSelect='none' m='1rem'>
          {introduction}
        </Text>
        <SubModuleList moduleId={id} submodules={submodules} />
      </Collapse>
    </Box>
  );
};

export default Module;