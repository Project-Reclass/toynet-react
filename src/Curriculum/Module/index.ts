import Module from './Module';
export * from './Module';
export * from './types';
export default Module;