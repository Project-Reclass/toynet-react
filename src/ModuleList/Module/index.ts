import Module from './Module';
export * from './Module';
export default Module;