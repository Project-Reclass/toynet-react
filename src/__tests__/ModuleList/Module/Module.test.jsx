import React from 'react';

import { renderTreeWithTheme } from 'src/common/test-utils/renderWithTheme';
import Module from 'src/Curriculum/Module';

const defaultProps = {
  title: 'First Module',
  progress: 8,
  type: 'parent',
  subModules: [
    {
      id: 1,
      moduleId: 1,
      title: 'First Article',
      progress: 42,
      type: 'quiz',
    }
  ]
}

describe('The Module', () => {
  it('should render and match snapshots', () => {
    const tree = renderTreeWithTheme(<Module {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});