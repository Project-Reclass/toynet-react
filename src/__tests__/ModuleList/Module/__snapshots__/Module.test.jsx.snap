// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Module should render and match snapshots 1`] = `
<div>
  <div
    className="css-1xeo39u"
  >
    <div
      className="css-11dqbsu"
      onClick={[Function]}
    >
      <svg
        className="css-1axzmpk"
        focusable={false}
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M12.8,5.4c-0.377-0.504-1.223-0.504-1.6,0l-9,12c-0.228,0.303-0.264,0.708-0.095,1.047 C2.275,18.786,2.621,19,3,19h18c0.379,0,0.725-0.214,0.895-0.553c0.169-0.339,0.133-0.744-0.095-1.047L12.8,5.4z"
          fill="currentColor"
        />
      </svg>
      <p
        className="css-grke7g"
      >
        First Module
      </p>
    </div>
  </div>
  <div
    aria-hidden={true}
    className="rah-static rah-static--height-zero css-vppuuz"
    style={
      Object {
        "height": 0,
        "overflow": "hidden",
      }
    }
  >
    <div
      style={
        Object {
          "WebkitTransition": "opacity 250ms ease 0ms",
          "opacity": 0,
          "transition": "opacity 250ms ease 0ms",
        }
      }
    >
      <div
        className="css-0"
      >
        <div
          className="css-1opj1b2"
        >
          <div
            className="css-k008qs"
          >
            <div
              className="css-j7qwjs"
            >
              <div
                className="css-1xc5jix"
                type="lock"
              >
                <svg
                  className="css-1qqmyhf"
                  focusable={false}
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19.5,9.5h-.75V6.75A6.751,6.751,0,0,0,5.533,4.811a1.25,1.25,0,1,0,2.395.717A4.251,4.251,0,0,1,16.25,6.75V9a.5.5,0,0,1-.5.5H4.5a2,2,0,0,0-2,2V22a2,2,0,0,0,2,2h15a2,2,0,0,0,2-2V11.5A2,2,0,0,0,19.5,9.5Zm-9.5,6a2,2,0,1,1,3,1.723V19.5a1,1,0,0,1-2,0V17.223A1.994,1.994,0,0,1,10,15.5Z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
            <div
              className="css-gmscv5"
            >
              <div
                className="css-gg4vpm"
              >
                <p
                  className="css-c02fy7"
                >
                  <a
                    className="css-q3xaa4"
                    href="/module/1/quiz/1"
                  >
                    First Article
                  </a>
                </p>
                <div
                  className="css-xtemij"
                >
                  <p
                    className="css-5yrbsi"
                  >
                    Module progress
                  </p>
                  <div
                    className="css-52gy4w"
                  >
                    <div
                      aria-valuemax={100}
                      aria-valuemin={0}
                      aria-valuenow={42}
                      className="css-d5agxk-Progress"
                      role="progressbar"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
